File: example_client.js
/**
 * Example JavaScript/TypeScript API Client for SAT Prep Reading Comprehension API
 * 
 * This is a reference implementation showing how to integrate with the API.
 * Adapt this to your specific framework (React, Vue, Angular, etc.)
 */

class SATPrepAPIClient {
  constructor(baseURL = 'http://localhost:8000/api/v1') {
    this.baseURL = baseURL;
  }

  /**
   * Helper method for making API requests
   */
  async request(endpoint, options = {}) {
    const url = `${this.baseURL}${endpoint}`;
    const config = {
      headers: {
        'Content-Type': 'application/json',
        'Accept': 'application/json',
        ...options.headers,
      },
      ...options,
    };

    if (config.body && typeof config.body === 'object') {
      config.body = JSON.stringify(config.body);
    }

    try {
      const response = await fetch(url, config);
      const data = await response.json();

      if (!response.ok) {
        throw new Error(data.error?.message || `HTTP ${response.status}`);
      }

      return data;
    } catch (error) {
      console.error('API Error:', error);
      throw error;
    }
  }

  // ==================== PASSAGES ====================

  /**
   * Get all passages
   * @param {Object} filters - Optional filters
   * @param {string} filters.difficulty - Filter by difficulty (Easy, Medium, Hard)
   * @param {number} filters.limit - Number of results
   * @param {number} filters.offset - Pagination offset
   * @returns {Promise<Object>} Passages list
   */
  async getPassages(filters = {}) {
    const params = new URLSearchParams();
    if (filters.difficulty) params.append('difficulty', filters.difficulty);
    if (filters.limit) params.append('limit', filters.limit);
    if (filters.offset) params.append('offset', filters.offset);

    const query = params.toString();
    return this.request(`/passages${query ? `?${query}` : ''}`);
  }

  /**
   * Get passage detail (includes correct answers - use with caution)
   * @param {string} passageId - Passage UUID
   * @returns {Promise<Object>} Passage with questions
   */
  async getPassageDetail(passageId) {
    return this.request(`/passages/${passageId}`);
  }

  /**
   * Get passage questions without correct answers (for active sessions)
   * @param {string} passageId - Passage UUID
   * @returns {Promise<Object>} Questions with options
   */
  async getPassageQuestions(passageId) {
    return this.request(`/passages/${passageId}/questions`);
  }

  // ==================== QUESTIONS ====================

  /**
   * Get question detail
   * @param {string} questionId - Question UUID
   * @returns {Promise<Object>} Question with options
   */
  async getQuestion(questionId) {
    return this.request(`/questions/${questionId}`);
  }

  // ==================== PROGRESS ====================

  /**
   * Get user progress summary
   * @returns {Promise<Object>} Progress summary
   */
  async getProgress() {
    return this.request('/progress');
  }

  /**
   * Get progress for a specific passage
   * @param {string} passageId - Passage UUID
   * @returns {Promise<Object>} Passage progress
   */
  async getPassageProgress(passageId) {
    return this.request(`/progress/passages/${passageId}`);
  }

  /**
   * Start a passage session
   * @param {string} passageId - Passage UUID
   * @param {string} startedAt - ISO 8601 timestamp (optional)
   * @returns {Promise<Object>} Session info
   */
  async startSession(passageId, startedAt = null) {
    return this.request(`/progress/passages/${passageId}/start`, {
      method: 'POST',
      body: {
        started_at: startedAt || new Date().toISOString(),
      },
    });
  }

  /**
   * Submit answers for a passage
   * @param {string} passageId - Passage UUID
   * @param {Array} answers - Array of {question_id, selected_option_index}
   * @param {number} timeSpentSeconds - Time spent in seconds
   * @returns {Promise<Object>} Submission results
   */
  async submitPassage(passageId, answers, timeSpentSeconds = 0) {
    return this.request(`/progress/passages/${passageId}/submit`, {
      method: 'POST',
      body: {
        answers,
        time_spent_seconds: timeSpentSeconds,
      },
    });
  }

  /**
   * Get review data for a completed passage
   * @param {string} passageId - Passage UUID
   * @returns {Promise<Object>} Review data with explanations
   */
  async getReview(passageId) {
    return this.request(`/progress/passages/${passageId}/review`);
  }

  // ==================== ANSWERS ====================

  /**
   * Submit a single answer (for real-time tracking)
   * @param {string} questionId - Question UUID
   * @param {number} selectedOptionIndex - 0-based index
   * @returns {Promise<Object>} Answer record
   */
  async submitAnswer(questionId, selectedOptionIndex) {
    return this.request('/answers', {
      method: 'POST',
      body: {
        question_id: questionId,
        selected_option_index: selectedOptionIndex,
      },
    });
  }

  /**
   * Get all answers for a passage
   * @param {string} passageId - Passage UUID
   * @returns {Promise<Object>} Answers array
   */
  async getPassageAnswers(passageId) {
    return this.request(`/answers/passage/${passageId}`);
  }
}

// ==================== USAGE EXAMPLES ====================

/**
 * Example: Complete passage flow
 */
async function exampleCompleteFlow() {
  const client = new SATPrepAPIClient();

  try {
    // 1. Get all passages
    const passagesResponse = await client.getPassages({ difficulty: 'Medium' });
    console.log('Available passages:', passagesResponse.results);

    if (passagesResponse.results.length === 0) {
      console.log('No passages available');
      return;
    }

    const passage = passagesResponse.results[0];
    const passageId = passage.id;

    // 2. Start a session (optional)
    const session = await client.startSession(passageId);
    console.log('Session started:', session.session_id);

    // 3. Get questions (without correct answers)
    const questionsData = await client.getPassageQuestions(passageId);
    console.log('Questions:', questionsData.questions);

    // 4. Simulate user answering questions
    const userAnswers = [];
    const startTime = Date.now();

    for (const question of questionsData.questions) {
      // Simulate user selecting an option (in real app, this comes from UI)
      const selectedIndex = Math.floor(Math.random() * question.options.length);
      userAnswers.push({
        question_id: question.id,
        selected_option_index: selectedIndex,
      });

      // Optionally submit answer in real-time
      // await client.submitAnswer(question.id, selectedIndex);
    }

    // 5. Calculate time spent
    const timeSpent = Math.floor((Date.now() - startTime) / 1000);

    // 6. Submit final answers
    const results = await client.submitPassage(passageId, userAnswers, timeSpent);
    console.log('Results:', {
      score: results.score,
      correct: results.correct_count,
      total: results.total_questions,
    });

    // 7. Get review with explanations
    const review = await client.getReview(passageId);
    console.log('Review data:', review);

  } catch (error) {
    console.error('Error in flow:', error);
  }
}

/**
 * Example: Track progress
 */
async function exampleTrackProgress() {
  const client = new SATPrepAPIClient();

  try {
    // Get user's overall progress
    const progress = await client.getProgress();
    console.log('Progress:', {
      completed: progress.completed_count,
      total: progress.total_passages,
      scores: progress.scores,
    });

    // Get progress for a specific passage
    const passageId = '550e8400-e29b-41d4-a716-446655440000';
    const passageProgress = await client.getPassageProgress(passageId);
    console.log('Passage progress:', passageProgress);

  } catch (error) {
    console.error('Error tracking progress:', error);
  }
}

/**
 * Example: React Hook usage
 */
// import { useState, useEffect } from 'react';
// 
// function usePassage(passageId) {
//   const [passage, setPassage] = useState(null);
//   const [loading, setLoading] = useState(true);
//   const [error, setError] = useState(null);
//   const client = new SATPrepAPIClient();
// 
//   useEffect(() => {
//     async function loadPassage() {
//       try {
//         setLoading(true);
//         const data = await client.getPassageQuestions(passageId);
//         setPassage(data);
//       } catch (err) {
//         setError(err);
//       } finally {
//         setLoading(false);
//       }
//     }
//     loadPassage();
//   }, [passageId]);
// 
//   return { passage, loading, error };
// }

// Export for use in other files
if (typeof module !== 'undefined' && module.exports) {
  module.exports = { SATPrepAPIClient };
}

// Export for ES6 modules
if (typeof window !== 'undefined') {
  window.SATPrepAPIClient = SATPrepAPIClient;
}

File: example_client.py
"""
Example Python API Client for SAT Prep Reading Comprehension API

This is a reference implementation showing how to integrate with the API from Python.
"""

import requests
from typing import Optional, List, Dict, Any
from datetime import datetime


class SATPrepAPIClient:
    """Client for interacting with the SAT Prep Reading Comprehension API"""

    def __init__(self, base_url: str = "http://localhost:8000/api/v1"):
        self.base_url = base_url
        self.session = requests.Session()
        self.session.headers.update({
            'Content-Type': 'application/json',
            'Accept': 'application/json',
        })

    def _request(
        self,
        method: str,
        endpoint: str,
        data: Optional[Dict] = None,
        params: Optional[Dict] = None
    ) -> Dict[str, Any]:
        """Make an API request"""
        url = f"{self.base_url}{endpoint}"
        
        try:
            response = self.session.request(
                method=method,
                url=url,
                json=data,
                params=params
            )
            response.raise_for_status()
            return response.json()
        except requests.exceptions.HTTPError as e:
            error_data = e.response.json() if e.response.content else {}
            raise Exception(
                error_data.get('error', {}).get('message', str(e))
            ) from e

    # ==================== PASSAGES ====================

    def get_passages(
        self,
        difficulty: Optional[str] = None,
        limit: Optional[int] = None,
        offset: Optional[int] = None
    ) -> Dict[str, Any]:
        """Get all passages with optional filtering"""
        params = {}
        if difficulty:
            params['difficulty'] = difficulty
        if limit:
            params['limit'] = limit
        if offset:
            params['offset'] = offset
        
        return self._request('GET', '/passages', params=params)

    def get_passage_detail(self, passage_id: str) -> Dict[str, Any]:
        """Get passage detail (includes correct answers - use with caution)"""
        return self._request('GET', f'/passages/{passage_id}')

    def get_passage_questions(self, passage_id: str) -> Dict[str, Any]:
        """Get passage questions without correct answers (for active sessions)"""
        return self._request('GET', f'/passages/{passage_id}/questions')

    # ==================== QUESTIONS ====================

    def get_question(self, question_id: str) -> Dict[str, Any]:
        """Get question detail"""
        return self._request('GET', f'/questions/{question_id}')

    # ==================== PROGRESS ====================

    def get_progress(self) -> Dict[str, Any]:
        """Get user progress summary"""
        return self._request('GET', '/progress')

    def get_passage_progress(self, passage_id: str) -> Dict[str, Any]:
        """Get progress for a specific passage"""
        return self._request('GET', f'/progress/passages/{passage_id}')

    def start_session(
        self,
        passage_id: str,
        started_at: Optional[str] = None
    ) -> Dict[str, Any]:
        """Start a passage session"""
        data = {
            'started_at': started_at or datetime.utcnow().isoformat() + 'Z'
        }
        return self._request('POST', f'/progress/passages/{passage_id}/start', data=data)

    def submit_passage(
        self,
        passage_id: str,
        answers: List[Dict[str, Any]],
        time_spent_seconds: int = 0
    ) -> Dict[str, Any]:
        """Submit answers for a passage"""
        data = {
            'answers': answers,
            'time_spent_seconds': time_spent_seconds
        }
        return self._request('POST', f'/progress/passages/{passage_id}/submit', data=data)

    def get_review(self, passage_id: str) -> Dict[str, Any]:
        """Get review data for a completed passage"""
        return self._request('GET', f'/progress/passages/{passage_id}/review')

    # ==================== ANSWERS ====================

    def submit_answer(
        self,
        question_id: str,
        selected_option_index: int
    ) -> Dict[str, Any]:
        """Submit a single answer (for real-time tracking)"""
        data = {
            'question_id': question_id,
            'selected_option_index': selected_option_index
        }
        return self._request('POST', '/answers', data=data)

    def get_passage_answers(self, passage_id: str) -> Dict[str, Any]:
        """Get all answers for a passage"""
        return self._request('GET', f'/answers/passage/{passage_id}')


# ==================== USAGE EXAMPLES ====================

def example_complete_flow():
    """Example: Complete passage flow"""
    client = SATPrepAPIClient()

    try:
        # 1. Get all passages
        passages_response = client.get_passages(difficulty='Medium')
        print(f"Available passages: {len(passages_response.get('results', []))}")

        if not passages_response.get('results'):
            print('No passages available')
            return

        passage = passages_response['results'][0]
        passage_id = passage['id']

        # 2. Start a session (optional)
        session = client.start_session(passage_id)
        print(f"Session started: {session['session_id']}")

        # 3. Get questions (without correct answers)
        questions_data = client.get_passage_questions(passage_id)
        questions = questions_data.get('questions', [])
        print(f"Found {len(questions)} questions")

        # 4. Simulate user answering questions
        import random
        user_answers = []
        start_time = datetime.now()

        for question in questions:
            # Simulate user selecting an option
            selected_index = random.randint(0, len(question['options']) - 1)
            user_answers.append({
                'question_id': question['id'],
                'selected_option_index': selected_index
            })

        # 5. Calculate time spent
        time_spent = int((datetime.now() - start_time).total_seconds())

        # 6. Submit final answers
        results = client.submit_passage(passage_id, user_answers, time_spent)
        print(f"Results: Score={results['score']}%, "
              f"Correct={results['correct_count']}/{results['total_questions']}")

        # 7. Get review with explanations
        review = client.get_review(passage_id)
        print(f"Review data retrieved: {len(review.get('answers', []))} answers")

    except Exception as e:
        print(f"Error in flow: {e}")


def example_track_progress():
    """Example: Track progress"""
    client = SATPrepAPIClient()

    try:
        # Get user's overall progress
        progress = client.get_progress()
        print(f"Progress: {progress['completed_count']}/{progress['total_passages']} completed")
        print(f"Scores: {progress['scores']}")

    except Exception as e:
        print(f"Error tracking progress: {e}")


if __name__ == '__main__':
    # Run examples
    print("=== Complete Flow Example ===")
    example_complete_flow()
    
    print("\n=== Progress Tracking Example ===")
    example_track_progress()


